src/kernels/ipywidgets/commonMessageCoordinator.ts

```diff
@@ -106,10 +115,45 @@ export class CommonMessageCoordinator {
                     this.onMessage(m.type, m.payload);
-
+                    const kernel = this.kernelProvider.get(this.document.uri);
                     // Special case the WidgetManager loaded message. It means we're ready
                     // to use a kernel. (IPyWidget Dispatcher uses this too)
                     if (m.type === IPyWidgetMessages.IPyWidgets_Ready) {
+                        if (kernel?.kernelConnectionMetadata.kind === 'startUsingRemoteKernelSpec') {
+                            traceInfoIfCI(
+                                'Web view is not ready to receive widget messages (kernel points to remote kernel spec)'
+                            );
+                            return;
+                        }
+                        if (!webview.isReady) {
+                            traceInfoIfCI('Web view is not ready to receive widget messages');
+                            return;
+                        }
+                        traceInfoIfCI('Web view is ready to receive widget messages');
+                        this.readyMessageReceived = true;
+                        this.sendPendingWebViewMessages();
+                        // At this point, we know the kernels are ready, and the webview is ready to receive messages.
+                        // Its possible the webview was initially unable to render some widgets, but now that everything is ready we
+                        // should be able to render them now.
+                        // E.g assume we're dealing with remote kernel specs,
+                        // Then we start a kernel, next we create a controller that points to the live kernel & change the controller.
+                        // What happens now is the, webview is re-loaded (as theres a change in the controllers).
+                        // However if the user were to run a cell, then the output would get displayed and it could end up,
+                        // being displayed in the old webview & when the webview is re-loaded, it would be displayed in the new webview.
+                        // However its possble the kernel is not yet ready at that point in time.
+                        // We could solve this issue easily by not executing cells, until the webview is ready,
+                        // however that would have significant performance implications.
+                        // Hence for ipywidgets, once the webview has completely been initialized, we can attempt to re-render the widgets.
+                        const nbEditor = this.serviceContainer
+                            .get<IVSCodeNotebook>(IVSCodeNotebook)
+                            .notebookEditors.find((item) => item.notebook === this.document);
+                        if (nbEditor) {
+                            traceInfoIfCI('Re-rendering widgets');
+                            notebooks
+                                .createRendererMessaging('jupyter-ipywidget-renderer')
```

Comment:
Should `jupyter-ipywidget-renderer` be elsewhere? Like in a `constants` kind of file?